lib/compress/zstd_opt.c

```diff
@@ -423,237 +424,241 @@ size_t ZSTD_compressBlock_opt_generic(ZSTD_CCtx* ctx,
+                        goto _shortestPath;
                     }
-                    best_off = i - (ip == anchor);
                     do {
-                        price = ZSTD_getPrice(optStatePtr, litlen, anchor, best_off, mlen - MINMATCH, ultra);
-                        if (mlen > last_pos || price < opt[mlen].price)
-                            SET_PRICE(mlen, mlen, i, litlen, price);   /* note : macro modifies last_pos */
-                        mlen--;
-                    } while (mlen >= minMatch);
+                        U32 const repPrice = ZSTD_getPrice(optStatePtr, initLL, anchor, repCode - ll0, repLen, ultra);
+                        if (repLen > last_pos || repPrice < opt[repLen].price)
+                            SET_PRICE(repLen, repLen, repCode, initLL, repPrice);   /* note : macro modifies last_pos */
+                        repLen--;
+                    } while (repLen >= minMatch);
         }   }   }
 
-        match_num = ZSTD_BtGetAllMatches_selectMLS(ctx, ip, iend, maxSearches, mls, matches, minMatch);
+        {   U32 const nb_matches = ZSTD_BtGetAllMatches_selectMLS(ctx, ip, iend, maxSearches, mls, matches, minMatch);
 
-        if (!last_pos && !match_num) { ip++; continue; }
+            if (!last_pos /*no repCode*/ && !nb_matches /*no match*/) { ip++; continue; }
 
-        if (match_num && (matches[match_num-1].len > sufficient_len || matches[match_num-1].len >= ZSTD_OPT_NUM)) {
-            best_mlen = matches[match_num-1].len;
-            best_off = matches[match_num-1].off;
-            cur = 0;
-            last_pos = 1;
-            goto _storeSequence;
-        }
+            if (nb_matches && (matches[nb_matches-1].len > sufficient_len)) {
+                /* large match => immediate encoding */
+                best_mlen = matches[nb_matches-1].len;
+                best_off = matches[nb_matches-1].off;
+                cur = 0;
+                last_pos = 1;
+                goto _shortestPath;
+            }
 
-        /* set prices using matches at position = 0 */
-        best_mlen = (last_pos) ? last_pos : minMatch;
-        for (u = 0; u < match_num; u++) {
-            mlen = (u>0) ? matches[u-1].len+1 : best_mlen;
-            best_mlen = matches[u].len;
-            while (mlen <= best_mlen) {
-                price = ZSTD_getPrice(optStatePtr, litlen, anchor, matches[u].off-1, mlen - MINMATCH, ultra);
-                if (mlen > last_pos || price < opt[mlen].price)
-                    SET_PRICE(mlen, mlen, matches[u].off, litlen, price);   /* note : macro modifies last_pos */
-                mlen++;
-        }   }
+            /* set prices for first matches from position == 0 */
+            {   U32 matchNb;
+                U32 pos = last_pos /*some repCode (assumed cheaper)*/ ? last_pos : minMatch;
+                for (matchNb = 0; matchNb < nb_matches; matchNb++) {
+                    U32 const end = matches[matchNb].len;
+                    while (pos <= end) {
```

Comment:
We are losing some optimality here, since matches aren't fully sorted by length (also different offsets). I guess we should leave it for another diff, since this one should have no functionality changes.

Reply:
Not sure to follow this one.
Matches are sorted by the matchfinder.
They should be sorted in both gradually longer length and longer offset.

In next patch, since repcodes and normal offsets match search are fused, normal offsets match lengths are necessarily larger than repcode ones.

Reply:
I believe they are still sorted.

For `minmatch==3`, it seems there is only a single attempt, using a dedicated hash table. So there can be only a single match, though its length might be accidentally larger than `3`.
Either way, such match modifies `bestLength`, which means future matches will only be added to the list if they are [larger than `bestLength`](https://github.com/facebook/zstd/blob/72c0f4574cfef1d5d5f35fc7277ee353793956ab/lib/compress/zstd_opt.c#L318).

So I believe they are still sorted in both gradually larger length and offset.